lib/model/topics.dart

```diff
@@ -0,0 +1,157 @@
+import 'dart:async';
+
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import '../api/model/model.dart';
+import '../api/route/channels.dart';
+import 'channel.dart';
+import 'store.dart';
+
+// similar to _apiSendMessage in lib/model/message.dart
+final _apiGetChannelTopics = getChannelTopics;
+
+/// The view-model for tracking channel topics.
+///
+/// Use [fetchChannelTopics] to first fetch the topics for a channel from the
+/// server and then use [getChannelTopics] to get those topics, affected
+/// by the relevant events.
+class Topics extends PerAccountStoreBase with ChangeNotifier {
+  Topics({required super.core});
+
+  /// Maps indexed by channel IDs, of the known latest message IDs in each topic.
+  ///
+  /// For example: `_latestMessageIdsByChannelTopic[channel.streamId][topic] = maxId`
+  ///
+  /// Occasionally, the latest message ID of a topic will refer to a message
+  /// that doesn't exist or is no longer in the topic.
+  /// This happens when the topic's latest message is deleted or moved
+  /// and we don't have enough information to replace it accurately.
+  /// (We don't keep a snapshot of all messages.)
+  // TODO(#2004): handle more cases where this can change
+  final Map<int, TopicKeyedMap<int>> _latestMessageIdsByChannelTopic = {};
+
+  final Map<int, Future<GetChannelTopicsResult>> _channelTopicsFetching = {};
+
+  /// Fetch topics in a channel from the server, only if they're not fetched yet.
+  ///
+  /// Once fetched, the data will be updated by events;
+  /// use [getChannelTopics] to consume the data.
+  Future<void> fetchChannelTopics(int channelId) async {
+    if (_latestMessageIdsByChannelTopic[channelId] != null) return;
+
+    Future<GetChannelTopicsResult>? future = _channelTopicsFetching[channelId];
+    // If another call has already started fetching topics for this channel,
+    // ignore this call.
+    if (future != null) return;
+
+    future = _apiGetChannelTopics(connection, channelId: channelId,
+      allowEmptyTopicName: true);
+    _channelTopicsFetching[channelId] = future;
+
+    try {
+      final result = await future;
+      (_latestMessageIdsByChannelTopic[channelId] = makeTopicKeyedMap())
+        .addEntries(result.topics.map((entry) => .new(entry.name, entry.maxId)));
```

Comment:
Let's add an `assert(_latestMessageIdsByChannelTopic[channelId] == null)` just above this.